fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/CTEProducerRewrite.java

```diff
@@ -0,0 +1,122 @@
+
+import org.apache.doris.nereids.CascadesContext;
+import org.apache.doris.nereids.jobs.batch.NereidsRewriter;
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.rules.rewrite.OneRewriteRuleFactory;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalCTEConsumer;
+import org.apache.doris.nereids.trees.plans.logical.LogicalCTEProducer;
+import org.apache.doris.nereids.trees.plans.logical.LogicalFilter;
+import org.apache.doris.nereids.trees.plans.logical.LogicalPlan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalProject;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import org.apache.commons.collections.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ *  Rewrite CTE Producer recursively.
+ */
+public class CTEProducerRewrite extends OneRewriteRuleFactory {
+
+    @Override
+    public Rule build() {
+        return logicalCTEProducer().when(p -> !p.isRewritten()).thenApply(ctx -> {
+            LogicalCTEProducer<? extends Plan> cteProducer = ctx.root;
+            Set<Expression> projects = ctx.cascadesContext.findProjectForProducer(cteProducer.getCteId());
+            LogicalPlan child = tryToConstructFilter(ctx.cascadesContext, cteProducer.getCteId(),
+                    (LogicalPlan) ctx.root.child());
+            if (CollectionUtils.isNotEmpty(projects)
+                    && ctx.cascadesContext.couldPruneColumnOnProducer(cteProducer.getCteId())) {
+                child = new LogicalProject(ImmutableList.copyOf(projects), child);
+            }
+            CascadesContext rewrittenCtx = ctx.cascadesContext.forkForCTEProducer(child);
+            NereidsRewriter rewriter = new NereidsRewriter(rewrittenCtx);
+            rewriter.execute();
+            return cteProducer.withChildrenAndProjects(ImmutableList.of(rewrittenCtx.getRewritePlan()),
+                    new ArrayList<>(child.getOutput()), true);
+        }).toRule(RuleType.CTE_PRODUCER_REWRITE);
+    }
+
+    /*
+     * An expression can only be pushed down if it has filter expressions on all consumers that reference the slot.
+     * For example, let's assume a producer has two consumers, consumer1 and consumer2:
+     *
+     * filter(a > 5 and b < 1) -> consumer1
+     * filter(a < 8) -> consumer2
+     *
+     * In this case, the only expression that can be pushed down to the producer is filter(a > 5 or a < 8).
+     */
+    private LogicalPlan tryToConstructFilter(CascadesContext cascadesContext, int cteId, LogicalPlan child) {
```

Comment:
just push `(a > 5 and b < 1) or (a < 8)` is simple and better

Reply:
Ok, I'll do it later